NullPointerException when entity primary key has the type `byte[]`

### Describe the bug

Quarkus crashes on startup on a `NullPointerException` when a JPA entity has a primary key of type `byte[]`.

```java
@Entity
public class Book
{
	@Id
	private byte[] id;
}
````

The `NullPointerException` does not appear when using standalone Hibernate, which leads me to believe this is a bug in Quarkus.

However, the problem can be fixed simply by bumping `hibernate-core` version to `6.2.8.Final` in pom.xml, which leads me to believe this might not be a Quarkus bug after all. Feel free to close the issue if that's the case.

### Expected behavior

Quarkus starts normally.

### Actual behavior

Crash on startup with the following stack trace:

```
java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.runner.bootstrap.StartupActionImpl$1.run(StartupActionImpl.java:104)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.RuntimeException: jakarta.persistence.PersistenceException: [PersistenceUnit: <default>] Unable to build Hibernate SessionFactory
        at io.quarkus.hibernate.orm.runtime.JPAConfig.startAll(JPAConfig.java:78)
        at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder.startAllPersistenceUnits(HibernateOrmRecorder.java:101)
        at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits1868654632.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.HibernateOrmProcessor$startPersistenceUnits1868654632.deploy(Unknown Source)
        ... 13 more
Caused by: jakarta.persistence.PersistenceException: [PersistenceUnit: <default>] Unable to build Hibernate SessionFactory
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.persistenceException(FastBootEntityManagerFactoryBuilder.java:126)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:86)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:74)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:80)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:55)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:156)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:64)
        ... 1 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.hibernate.metamodel.mapping.JdbcMapping.getJdbcType()" because the return value of "org.hibernate.metamodel.mapping.JdbcMappingContainer.getJdbcMapping(int)" is null
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.renderParameterAsParameter(AbstractSqlAstTranslator.java:6240)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitParameterAsParameter(AbstractSqlAstTranslator.java:6234)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitParameter(AbstractSqlAstTranslator.java:6227)
        at org.hibernate.sql.exec.internal.AbstractJdbcParameter.accept(AbstractJdbcParameter.java:57)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitInArrayPredicate(AbstractSqlAstTranslator.java:6952)
        at org.hibernate.sql.ast.tree.predicate.InArrayPredicate.accept(InArrayPredicate.java:41)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitWhereClause(AbstractSqlAstTranslator.java:3087)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitQuerySpec(AbstractSqlAstTranslator.java:3033)
        at org.hibernate.sql.ast.tree.select.QuerySpec.accept(QuerySpec.java:119)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.visitSelectStatement(AbstractSqlAstTranslator.java:986)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.translateSelect(AbstractSqlAstTranslator.java:849)
        at org.hibernate.sql.ast.spi.AbstractSqlAstTranslator.translate(AbstractSqlAstTranslator.java:799)
        at org.hibernate.loader.ast.internal.EntityBatchLoaderArrayParam.prepare(EntityBatchLoaderArrayParam.java:189)
        at org.hibernate.persister.entity.AbstractEntityPersister.prepareLoader(AbstractEntityPersister.java:3447)
        at org.hibernate.persister.entity.AbstractEntityPersister.postInstantiate(AbstractEntityPersister.java:3441)
        at org.hibernate.metamodel.model.domain.internal.MappingMetamodelImpl.finishInitialization(MappingMetamodelImpl.java:204)
        at org.hibernate.internal.SessionFactoryImpl.initializeMappingModel(SessionFactoryImpl.java:320)
        at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:270)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:84)
        ... 6 more
```

### How to Reproduce?

Here is a minimal reproducing project: [byte-array-primary-key-issue.zip](https://github.com/quarkusio/quarkus/files/12580270/byte-array-primary-key-issue.zip)

Reproduce the issue by running `mvn quarkus:dev` or by running the test QuarkusBookResourceTest.

Contrast this with running `mvn test -Dtest="PureHibernateBookTest"`, which succeeds.

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.22000.2295]

### Output of `java -version`

java version "17.0.8" 2023-07-18 LTS

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.3.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)

### Additional information

This used to work at least in Quarkus version `2.16.9.Final`.

The `null` value causing the exception originates from [ArrayJavaType.java:78](https://github.com/hibernate/hibernate-orm/blob/3c625ead057dd26a5d7d5050ae9d81bc0b48d819/hibernate-core/src/main/java/org/hibernate/type/descriptor/java/ArrayJavaType.java#L78).

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -99,10 +99,10 @@
         <classmate.version>1.5.1</classmate.version>
         <!-- When updating, align bytebuddy.version to Hibernate needs as well (just below),
              as well as hibernate-orm.version-for-documentation in docs/pom.xml -->
-        <hibernate-orm.version>6.2.7.Final</hibernate-orm.version>
+        <hibernate-orm.version>6.2.9.Final</hibernate-orm.version>
         <bytebuddy.version>1.14.7</bytebuddy.version> <!-- Version controlled by Hibernate ORM's needs -->
         <hibernate-commons-annotations.version>6.0.6.Final</hibernate-commons-annotations.version> <!-- version controlled by Hibernate ORM -->
-        <hibernate-reactive.version>2.0.4.Final</hibernate-reactive.version>
+        <hibernate-reactive.version>2.0.5.Final</hibernate-reactive.version>
         <hibernate-validator.version>8.0.1.Final</hibernate-validator.version>
         <!-- When updating, align hibernate-search.version-for-documentation in docs/pom.xml -->
         <hibernate-search.version>6.2.1.Final</hibernate-search.version>

